### ISSUE TITLE
IEntityTypeConfiguration.Configure() is not fired for entities without default constructor

### ISSUE THREAD

**Issue by apostnov** [dotnet/EntityFramework.Docs#3207]
## File a bug

I looked through [open](https://github.com/dotnet/efcore/issues) and [closed](https://github.com/dotnet/efcore/issues?q=is%3Aissue+is%3Aclosed) issues and haven't found any similar so here is my report:

For simple enity 
```
    public class User: IEntityTypeConfiguration<User>
    {
        public long Id { get; set; }

        public string Name { get; set; } = string.Empty;
        
        public void Configure(EntityTypeBuilder<User> user)
        {
            user.HasIndex(x => x.Name);
        }
    }
```
when I use IEntityTypeConfiguration to reverse DbContext configuration from context `OnModelCreating()` into entity itself it works fine generating migration having `migrationBuilder.CreateIndex()`

```
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.CreateTable(
                name: "user",
                columns: table => new
                {
                    id = table.Column<long>(type: "bigint", nullable: false)
                        .Annotation("Npgsql:ValueGenerationStrategy", NpgsqlValueGenerationStrategy.IdentityByDefaultColumn),
                    name = table.Column<string>(type: "text", nullable: false)
                },
                constraints: table =>
                {
                    table.PrimaryKey("pk_user", x => x.id);
                });

            **migrationBuilder.CreateIndex(
                name: "ix_user_name",
                table: "user",
                column: "name");**
        }
```

If I want to have custom constructor like this

```
    public class User : IEntityTypeConfiguration<User>
    {
        public User(string name)
        {
            Name = name;
        }

        public long Id { get; set; }

        public string Name { get; set; } = string.Empty;

        public void Configure(EntityTypeBuilder<User> user)
        {
            user.HasIndex(x => x.Name);
        }
    }
```

it generates empty migration unless `public DbSet<User> Users { get; set; } = null!;` is included into DbContext explicitly. Migration generated then is as follows:
```
protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.CreateTable(
                name: "users",
                columns: table => new
                {
                    id = table.Column<long>(type: "bigint", nullable: false)
                        .Annotation("Npgsql:ValueGenerationStrategy", NpgsqlValueGenerationStrategy.IdentityByDefaultColumn),
                    name = table.Column<string>(type: "text", nullable: false)
                },
                constraints: table =>
                {
                    table.PrimaryKey("pk_users", x => x.id);
                });
        }
```

which is obvious proof that `IEntityTypeConfiguration.Configure() ` is not fired during configuration.

_Expected behavior is to have the same migration no matter what constructors entity has cause DB model does not depend on them._

Ofcourse I can add default constructor explicitly, but how can I have custom one only e.g. for the purpose to have insuranse `Name` property is set ?

### Include provider and version information

EF Core version: 5.0.2
Database provider: Npgsql.EntityFrameworkCore.PostgreSQL
Target framework:  .NET 5.0)
Operating system: Windows 10
IDE:  Visual Studio 2019 16.8.6

**Comment by apostnov** [comment]
> @apostnov I'll move this to the docs repo to track documenting this.
> 
> If ApplyConfigurationsFromAssembly were to do this, then what value would it pass to the constructor?

My thought was if for some reason `ApplyConfigurationsFromAssembly()` needs entity to be instantiated with parameterless constructor then it could use Activator to call non-public constructor like `Foo foo = (Foo)Activator.CreateInstance(typeof(Foo), true);` if it exists and there is no public one. @ajcvickers

**Comment by ajcvickers** [comment]
@apostnov This would only work if there is a non-public parameterless constructor that could be called. (Have you verified that this doesn't work already?)

I think the main problem here is that you are are making the entity type itself implement `IEntityTypeConfiguration`, which is unusual and results in this kind of problematic coupling.

**Comment by apostnov** [comment]
@ajcvickers Yes, I have verified that having protected parameterless constructor and parametrised public one does not solve issue.

**Comment by mrm** [comment]
> If ApplyConfigurationsFromAssembly were to do this, then what value would it pass to the constructor?

I did expect this to work with DI though (using the `IEntityTypeConfiguration<TEntity>` class just for configuration). How would you use DI in that case?

**Comment by Aykuttonpc** [comment]
Hi! I'd like to take a stab at this. Looking at the current implementation of `ModelBuilder.ApplyConfigurationsFromAssembly`, the actual behavior is slightly more nuanced than the original discussion suggests:

- A parameterless constructor is required, but it may be **public or non-public** — the implementation uses `BindingFlags.Public | BindingFlags.NonPublic` and `Activator.CreateInstance(type, nonPublic: true)`.
- Types without a parameterless constructor are **not silently skipped** — a `SkippedEntityTypeConfigurationWarning` is logged.

I'll add a `[!NOTE]` block in `entity-framework/core/modeling/index.md` (next to the existing "order is undefined" note) capturing this. Does this sound good?
